retrieval/discovery/ec2.go

```diff
@@ -110,12 +115,25 @@ func (ed *EC2Discovery) refresh() (*config.TargetGroup, error) {
 				labels := model.LabelSet{
 					ec2LabelInstanceID: model.LabelValue(*inst.InstanceId),
 				}
+				labels[ec2LabelPrivateIP] = model.LabelValue(*inst.PrivateIpAddress)
+				addr := fmt.Sprintf("%s:%d", *inst.PrivateIpAddress, ed.port)
+
 				if inst.PublicIpAddress != nil {
 					labels[ec2LabelPublicIP] = model.LabelValue(*inst.PublicIpAddress)
+					labels[ec2LabelPublicDns] = model.LabelValue(*inst.PublicDnsName)
+					// replace address with public dns name if present
+					addr = fmt.Sprintf("%s:%d", *inst.PublicDnsName, ed.port)
```

Comment:
If the user assign a public ip to the instance, he/she may want to use that for connections, as we don't know the user's intention I think that it's fine to leave the prometheus server's location (inside or outside the instance's network) be decisive via dns resolution as dns in a ec2 or vpc network uses split-horizon based on the client's location.

The idea behind that change  is that service discovery should return meaninful information for the client's perspective, so if the Prometheus server is running on the user's laptop, in the ec2 network or in a vpc, the user will not have to change his scrape config as the information will be valid regardless. 

Still I understand your concern about the private ip convention, would a flag in ec2sdconfig [here](https://github.com/federicobaldo/prometheus/blob/ec2_sd_improvements/config/config.go#L681), for instance "use_hostname: true", be ok?